### ISSUE TITLE
Non-existing page redirects to no-cache=1 and not to 404 error page

### ISSUE THREAD

**Issue by wiched** [gatsbyjs/gatsby#8169]
Hello, 

On Gatsby develop non-existing urls redirect to the 404 error page, but on the build, they all redirect to the homepage with URL attribute /?no-cache=1

I tried to read the code for the load-directly-or-404.js but I didn't quite understand it.

How do fix that and make it redirect to 404?

Git repo: https://github.com/wiched/gatsby
Live url: https://www.prettystyle.store

**Comment by davidbailey00** [comment]
Hi @wiched, after upgrading your packages I can't reproduce this issue, so closing. Please see https://github.com/davidbailey00/prettystyle and https://nostalgic-bell-7077d9.netlify.com/ and let me know when you've sorted the main site so I can delete my copy of the repo :)

**Comment by SMerdzhanov** [comment]
Hey @davidbailey00,

I cannot agree with you. My configuration was working perfectly fine as well, but unfortunately, after updating all of my packages I face this same odd behavior... Actually I see more people experiencing this same behavior with fresh projects.

Using `"gatsby": "next",` @ `version "2.0.0-rc.25"`

**Comment by davidbailey00** [comment]
@SMerdzhanov Please can you create a reproduction repository so I can see the behaviour?

**Comment by robertoaikode** [comment]
I have the same problem! Any solution?

**Comment by DSchau** [comment]
I'd recommend updating your gatsby-* packages to their latest 2.x versions.

One really helpful tool is using the command `yarn upgrade-interactive` which will give you an intuitive way to see the latest versions and pick and choose what you want.

**Comment by davidbailey00** [comment]
@robertoaikode This isn't the same as the issue you opened - in this case it turned out that the problem with @wiched's site was an Apache misconfiguration. (cc @DSchau)

**Comment by SMerdzhanov** [comment]
Hey @davidbailey00, sorry for the belated response.

There's something odd happening. After cleaning up `node_modules`, `public` and `.cache`, using `yarn` everything seems fine. But I shared my project with a couple of people using `npm` and they all face issues with it.

I switched to `npm` to test and it turns out I also face some troubles with it. 

**I think this isn't the best spot to discuss, but throwing in just if someone else encounters it.**

The project I'm using is [mkit-bundle-gatsby](https://github.com/mkitio/mkit-bundle-gatsby).

```
// package.json sliced
...
"gatsby": "next",
"gatsby-plugin-google-analytics": "^1.0.31",
"gatsby-plugin-manifest": "next",
"gatsby-plugin-offline": "next",
"gatsby-plugin-react-helmet": "next",
"gatsby-plugin-robots-txt": "^1.3.0",
"gatsby-plugin-sitemap": "^1.2.25",
"gatsby-plugin-webpack-size": "0.0.2",
...
```

**Using `yarn`**
Everything seems to be fine and working as expected. 

Dev - non-existing-route -> http://localhost:8000/qweasdzxc -> Gatsby.js development 404 page

Prod -> non-existing route -> http://localhost:9000/qweasdzxc?no-cache=1

**Using `npm v6.4.1`**
Using the very same project, without touching any of its files has different behavior.

Running `build` & `serve` throws an odd error:

```
WebpackError: TypeError: uniqBy is not a function
```

The project doesn't use `lodash` anywhere, so I assume there's something wrong with deps and their resolution.

----

**My advise** is to do:
1. cleanup `node_modules`, `.cache` and `public`
2. do a fresh `npm install` and test it out
3. if you encounter any odd behavior or errors start again from step 1. but using `yarn`

Thank you for following up @davidbailey00 and sorry for the confusion.
